File: app/ios/Sources/CatLaserAuth/AttestationBinding.swift
import Foundation

/// Per-call freshness input mixed into the ECDSA signature of a
/// `DeviceAttestation`. Without this, the signed payload is just the
/// fingerprint hash — a value that is stable for a given device and
/// therefore indefinitely replayable by anyone who ever captures a valid
/// attestation header (server-side logs, a one-off TLS bypass, a future
/// pinning hook). Binding each signature to a context-specific value
/// collapses that window to a single request.
///
/// Six variants, one per authenticated endpoint:
///
/// * `.request(timestamp:)` — placed on the outbound `requestMagicLink`
///   call. Timestamp is Unix seconds at the moment the attestation is
///   built. Server rejects skew > ~60s; that implicitly limits the
///   replay window to one minute even if the full header leaks.
///
/// * `.verify(token:)` — placed on the `completeMagicLink` call. The
///   magic-link token is single-use and server-issued; binding the
///   signature to it means a request-time attestation cannot satisfy a
///   verify call, and a verify-time attestation cannot be re-used with
///   a different token. Request and verify signatures are disjoint by
///   construction even when every other input is identical.
///
/// * `.social(timestamp:rawNonce:)` — placed on the `exchangeSocial`
///   call for Apple / Google ID-token sign-in. Binds TWO freshness
///   inputs: the per-sign-in raw nonce (client-generated and also
///   echoed in the provider ID token's `nonce` claim — hashed for
///   Apple, verbatim for Google — so the server can three-way-match the
///   binding against the body and the provider claim), AND a wall-clock
///   Unix-seconds timestamp under the same ±60s skew contract as
///   `.request` / `.signOut`.
///
///   The timestamp is the v4 addition — without it a captured
///   `(body, attestation)` pair could be replayed for the full ID-token
///   lifetime (~10 minutes on Apple, up to an hour on Google) because
///   replay does not require the non-extractable Secure-Enclave private
///   key: the attacker re-submits the original bytes verbatim.
///   Enforcing the timestamp pins the replay window to ±60s, matching
///   the rest of the attestation surface.
///
/// * `.signOut(timestamp:)` — placed on the `signOut` call. Same
///   timestamp + skew contract as `.request`; distinct tag prevents a
///   captured `req:` signature from being submitted to the sign-out
///   endpoint (and vice versa). Device-binds the revocation so a leaked
///   bearer token alone cannot invalidate a session without also
///   forging a signature under the original Secure-Enclave key.
///
/// * `.api(timestamp:)` — placed on every authenticated API call after
///   sign-in. Same timestamp + skew contract as the other timestamped
///   bindings. Distinct tag prevents a captured sign-in-time attestation
///   from being replayed against a protected route, and vice versa. The
///   server-side protected-route middleware verifies every `api:`
///   binding under the pk captured at sign-in, so a leaked bearer alone
///   cannot act — a fresh SE signature is also required. ADR-006
///   documents the full contract.
///
/// * `.device(timestamp:)` — placed on the `AuthRequest` frame the app
///   sends as the FIRST frame on every TCP connection to the device
///   daemon. The frame carries a full v4 attestation header with
///   `bnd = "dev:<unix_seconds>"`. The device parses the header,
///   reconstructs the signed bytes, verifies the P-256 ECDSA signature
///   against the `pk` embedded in the frame, and then matches that pk
///   against its cached ACL of authorized users. Skew is checked by the
///   device against its local clock with the same ±60s window as the
///   server-side bindings. Distinct tag prevents any other captured
///   attestation from being replayed as a device auth frame, and
///   prevents a captured device frame from being replayed against a
///   coordination-server endpoint.
///
/// Wire format: the binding renders to a tagged UTF-8 string
/// (`"req:<ts>"`, `"ver:<token>"`, `"sis:<ts>:<rawNonce>"`,
/// `"out:<ts>"`, `"api:<ts>"`, or `"dev:<ts>"`) that is both:
///
/// 1. Placed on the wire under the `bnd` key of the attestation payload.
/// 2. Appended to the 32-byte `fph` hash and fed as the ECDSA message —
///    `sig = ECDSA-P256-SHA256(fph_raw || bnd_utf8, private_key)`.
///
/// The tagging prevents cross-context confusion: a server that strips the
/// prefix before parsing would still refuse to accept a `ver:` signature
/// as a `req:` input (or a `sis:` signature as either), because the raw
/// signed bytes differ in their first four characters.
public enum AttestationBinding: Sendable, Equatable {
    case request(timestamp: Int64)
    case verify(token: String)
    case social(timestamp: Int64, rawNonce: String)
    case signOut(timestamp: Int64)
    case api(timestamp: Int64)
    case device(timestamp: Int64)

    /// Upper bound on the UTF-8 size of `wireBytes`. Magic-link tokens and
    /// raw nonces are small (tens of bytes); this bound exists so a
    /// malformed caller cannot inflate the attestation header. Enforced
    /// by `DeviceAttestationEncoder` before placing the binding on the
    /// wire.
    public static let maxWireBytes = 1024

    /// Canonical UTF-8 encoding placed on the wire under `bnd` and mixed
    /// into the ECDSA signature input.
    public var wireValue: String {
        switch self {
        case let .request(timestamp): "req:\(timestamp)"
        case let .verify(token): "ver:\(token)"
        case let .social(timestamp, rawNonce): "sis:\(timestamp):\(rawNonce)"
        case let .signOut(timestamp): "out:\(timestamp)"
        case let .api(timestamp): "api:\(timestamp)"
        case let .device(timestamp): "dev:\(timestamp)"
        }
    }

    public var wireBytes: Data {
        Data(wireValue.utf8)
    }

    /// Parse a wire value back into a binding. Used by the server plugin's
    /// Swift port and by tests; the app itself only ever encodes.
    ///
    /// Tolerates nothing: unknown tag, empty payload, leading-zero /
    /// signed / non-numeric timestamps, and control characters in the
    /// token or nonce all reject.
    public static func decode(wireValue: String) throws(AuthError) -> AttestationBinding {
        guard wireValue.utf8.count <= maxWireBytes else {
            throw .attestationFailed("bnd exceeds \(maxWireBytes) bytes (got \(wireValue.utf8.count))")
        }
        if let ts = wireValue.prefixStrippedIfMatches("req:") {
            guard let parsed = Int64(ts), parsed > 0, String(parsed) == ts else {
                throw .attestationFailed("bnd timestamp is not a positive decimal Int64")
            }
            return .request(timestamp: parsed)
        }
        if let token = wireValue.prefixStrippedIfMatches("ver:") {
            guard !token.isEmpty else {
                throw .attestationFailed("bnd verify token is empty")
            }
            let disallowed = CharacterSet.whitespacesAndNewlines.union(.controlCharacters)
            if token.unicodeScalars.contains(where: disallowed.contains) {
                throw .attestationFailed("bnd verify token contains control characters")
            }
            return .verify(token: token)
        }
        if let payload = wireValue.prefixStrippedIfMatches("sis:") {
            // `sis:` carries `<timestamp>:<rawNonce>`. Split on the FIRST
            // `:` so a future widening of the nonce charset that happened
            // to contain `:` still round-trips — iOS `NonceGenerator`
            // currently emits base64url-no-pad (no `:` possible), so this
            // is strictly forward-compatibility insurance.
            guard let sep = payload.firstIndex(of: ":") else {
                throw .attestationFailed(
                    "bnd sis: missing ':<rawNonce>' suffix (expected 'sis:<unix_seconds>:<raw_nonce>')",
                )
            }
            let tsString = String(payload[..<sep])
            let rawNonce = String(payload[payload.index(after: sep)...])
            guard let parsed = Int64(tsString), parsed > 0, String(parsed) == tsString else {
                throw .attestationFailed("bnd sis: timestamp is not a positive decimal Int64")
            }
            guard !rawNonce.isEmpty else {
                throw .attestationFailed("bnd social raw nonce is empty")
            }
            let disallowed = CharacterSet.whitespacesAndNewlines.union(.controlCharacters)
            if rawNonce.unicodeScalars.contains(where: disallowed.contains) {
                throw .attestationFailed("bnd social raw nonce contains control characters")
            }
            return .social(timestamp: parsed, rawNonce: rawNonce)
        }
        if let ts = wireValue.prefixStrippedIfMatches("out:") {
            guard let parsed = Int64(ts), parsed > 0, String(parsed) == ts else {
                throw .attestationFailed("bnd sign-out timestamp is not a positive decimal Int64")
            }
            return .signOut(timestamp: parsed)
        }
        if let ts = wireValue.prefixStrippedIfMatches("api:") {
            guard let parsed = Int64(ts), parsed > 0, String(parsed) == ts else {
                throw .attestationFailed("bnd api timestamp is not a positive decimal Int64")
            }
            return .api(timestamp: parsed)
        }
        if let ts = wireValue.prefixStrippedIfMatches("dev:") {
            guard let parsed = Int64(ts), parsed > 0, String(parsed) == ts else {
                throw .attestationFailed("bnd device timestamp is not a positive decimal Int64")
            }
            return .device(timestamp: parsed)
        }
        throw .attestationFailed(
            "bnd has no recognised tag (expected 'req:', 'ver:', 'sis:', 'out:', 'api:', or 'dev:')",
        )
    }
}

private extension String {
    func prefixStrippedIfMatches(_ prefix: String) -> String? {
        guard hasPrefix(prefix) else { return nil }
        return String(dropFirst(prefix.count))
    }
}

File: server/src/lib/attestation-binding.ts
/**
 * Server-side port of the iOS `AttestationBinding` wire format.
 *
 * `bnd` is the per-call freshness input mixed into the ECDSA signature of a
 * `DeviceAttestation`. It renders on the wire as one of four tagged UTF-8
 * strings, distinguished by their first four characters:
 *
 * - `req:<unix_seconds>` — outbound magic-link request. Timestamp-skewed.
 * - `ver:<magic_link_token>` — magic-link verify. Bound to the server-issued token.
 * - `sis:<raw_nonce>` — social ID-token exchange. Bound to the per-sign-in nonce
 *   which is also echoed into the provider's ID token (verbatim for Google,
 *   hashed for Apple) and into the request body's `idToken.nonce` field for a
 *   three-way match.
 * - `out:<unix_seconds>` — sign-out. Same timestamp/skew contract as `req:`,
 *   distinct tag prevents a captured `req:` header from being replayed against
 *   sign-out (and vice versa).
 * - `api:<unix_seconds>` — every authenticated API call after sign-in. Same
 *   timestamp/skew contract as `req:` and `out:`; distinct tag prevents a
 *   captured sign-in-time attestation from being replayed against a protected
 *   route, and vice versa. Step 7 mounts this binding on the protected-route
 *   middleware; step 6 establishes the parser, encoder, and skew semantics so
 *   the crypto floor is consistent across all five bindings.
 *
 * The tagged prefix also prevents cross-context confusion: a caller that
 * stripped the prefix before signing/verifying would still see the raw signed
 * bytes differ in their first four characters between contexts.
 */

/**
 * Upper bound on the UTF-8 size of the wire value. Magic-link tokens and raw
 * nonces are small (tens of bytes); this bound exists so a malformed client
 * cannot inflate the attestation header. Enforced here at parse time and
 * mirrored on the iOS encode side.
 */
export const MAX_BND_WIRE_BYTES = 1024;

export type AttestationBinding =
  | { readonly tag: 'request'; readonly timestamp: bigint }
  | { readonly tag: 'verify'; readonly token: string }
  | { readonly tag: 'social'; readonly rawNonce: string }
  | { readonly tag: 'signOut'; readonly timestamp: bigint }
  | { readonly tag: 'api'; readonly timestamp: bigint };

export type AttestationBindingTag = AttestationBinding['tag'];

export type AttestationParseCode =
  | 'ATTESTATION_BND_TOO_LARGE'
  | 'ATTESTATION_BND_UNKNOWN_TAG'
  | 'ATTESTATION_BND_BAD_TIMESTAMP'
  | 'ATTESTATION_BND_EMPTY_TOKEN'
  | 'ATTESTATION_BND_CONTROL_CHARS';

export class AttestationParseError extends Error {
  public readonly code: AttestationParseCode;

  public constructor(code: AttestationParseCode, message: string) {
    super(message);
    this.name = 'AttestationParseError';
    this.code = code;
  }
}

const TAG_REQUEST = 'req:';
const TAG_VERIFY = 'ver:';
const TAG_SOCIAL = 'sis:';
const TAG_SIGN_OUT = 'out:';
const TAG_API = 'api:';

const DECIMAL_PATTERN = /^(?:0|[1-9]\d*)$/v;
const INT64_MAX = 9_223_372_036_854_775_807n;
const INT64_MAX_DECIMAL_DIGITS = 19;

/**
 * Any Unicode whitespace character OR Unicode control/format character.
 * `\p{Cc}` covers C0 + DEL + C1; `\p{Cf}` covers formatting marks such as
 * ZWNBSP (U+FEFF) and the bidi overrides; `\p{White_Space}` covers ASCII
 * space, NBSP, line/paragraph separators, the ogham space mark, etc.
 * Unicode property escapes require the `v` flag. This mirrors iOS's
 * `CharacterSet.whitespacesAndNewlines.union(.controlCharacters)` set,
 * which itself spans Unicode General Categories Cc + Cf plus whitespace:
 * a string iOS would refuse to transmit must also be rejected here, and
 * vice versa.
 */
const DISALLOWED_PAYLOAD_CHAR = /[\p{Cc}\p{Cf}\p{White_Space}]/v;

const utf8Encoder = new TextEncoder();
const utf8ByteLength = (value: string): number => utf8Encoder.encode(value).length;

/** Numeric payload after `req:` / `out:`. Positive, decimal, no leading zeros. */
const parseTimestamp = (input: string): bigint => {
  if (input.length === 0) {
    throw new AttestationParseError('ATTESTATION_BND_BAD_TIMESTAMP', 'bnd timestamp is empty');
  }
  if (!DECIMAL_PATTERN.test(input)) {
    throw new AttestationParseError(
      'ATTESTATION_BND_BAD_TIMESTAMP',
      'bnd timestamp is not a positive decimal integer',
    );
  }
  // Length cap under 2^63-1's 19-digit ceiling avoids BigInt overflow noise;
  // anything beyond that is certainly bogus.
  if (input.length > INT64_MAX_DECIMAL_DIGITS) {
    throw new AttestationParseError(
      'ATTESTATION_BND_BAD_TIMESTAMP',
      `bnd timestamp exceeds Int64 decimal width (${INT64_MAX_DECIMAL_DIGITS.toString()} digits)`,
    );
  }
  const parsed = BigInt(input);
  if (parsed <= 0n || parsed > INT64_MAX) {
    throw new AttestationParseError(
      'ATTESTATION_BND_BAD_TIMESTAMP',
      `bnd timestamp is not in positive Int64 range (got ${parsed.toString()})`,
    );
  }
  return parsed;
};

const parseOpaqueToken = (input: string, kind: string): string => {
  if (input.length === 0) {
    throw new AttestationParseError('ATTESTATION_BND_EMPTY_TOKEN', `bnd ${kind} is empty`);
  }
  if (DISALLOWED_PAYLOAD_CHAR.test(input)) {
    throw new AttestationParseError(
      'ATTESTATION_BND_CONTROL_CHARS',
      `bnd ${kind} contains control or whitespace characters`,
    );
  }
  return input;
};

/**
 * Decode a `bnd` wire value into a strongly-typed binding.
 *
 * Tolerates nothing beyond the exact wire format. Unknown tags, empty
 * payloads, timestamps with leading zeros, signed timestamps, non-decimal
 * timestamps, timestamps ≤ 0, and control characters in tokens/nonces all
 * reject. Mirrors the iOS `AttestationBinding.decode` contract byte-for-byte.
 */
export const decodeAttestationBinding = (wireValue: string): AttestationBinding => {
  const byteLength = utf8ByteLength(wireValue);
  if (byteLength > MAX_BND_WIRE_BYTES) {
    throw new AttestationParseError(
      'ATTESTATION_BND_TOO_LARGE',
      `bnd exceeds ${MAX_BND_WIRE_BYTES.toString()} bytes (got ${byteLength.toString()})`,
    );
  }

  if (wireValue.startsWith(TAG_REQUEST)) {
    return { tag: 'request', timestamp: parseTimestamp(wireValue.slice(TAG_REQUEST.length)) };
  }
  if (wireValue.startsWith(TAG_VERIFY)) {
    return {
      tag: 'verify',
      token: parseOpaqueToken(wireValue.slice(TAG_VERIFY.length), 'verify token'),
    };
  }
  if (wireValue.startsWith(TAG_SOCIAL)) {
    return {
      tag: 'social',
      rawNonce: parseOpaqueToken(wireValue.slice(TAG_SOCIAL.length), 'social raw nonce'),
    };
  }
  if (wireValue.startsWith(TAG_SIGN_OUT)) {
    return { tag: 'signOut', timestamp: parseTimestamp(wireValue.slice(TAG_SIGN_OUT.length)) };
  }
  if (wireValue.startsWith(TAG_API)) {
    return { tag: 'api', timestamp: parseTimestamp(wireValue.slice(TAG_API.length)) };
  }
  throw new AttestationParseError(
    'ATTESTATION_BND_UNKNOWN_TAG',
    `bnd has no recognised tag (expected '${TAG_REQUEST}', '${TAG_VERIFY}', '${TAG_SOCIAL}', '${TAG_SIGN_OUT}', or '${TAG_API}')`,
  );
};
